cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -430,7 +433,33 @@ template <>
 hash_value_type CUDA_DEVICE_CALLABLE
 SparkMurmurHash3_32<numeric::decimal128>::operator()(numeric::decimal128 const& key) const
 {
-  return this->compute<__int128_t>(key.value());
+  // Generates the Spark murmur3 hash value, mimicing the conversion
+  // java.math.BigDecimal.valueOf(unscaled_value, _scale).unscaledValue().toByteArray()
+  // https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L381
+  __int128_t const val    = key.value();
+  int8_t const* data      = reinterpret_cast<int8_t const*>(&val);
+  int8_t const sign_bit   = data[15] & 0x80;
+  int8_t const zero_value = sign_bit ? 0xff : 0x00;
+  int32_t length          = 15;
+
+  // Search for first non-zero byte in the unscaled value, shortening the hashed data
```

Comment:
Do we really mean the *last* non-zero byte? This depends on the endianness of the representation, right? If so please update the variable names in other suggestions.

Reply:
```
  // Searching from the big end, find the first non-zero byte in the unscaled little endian value
  // The zero bytes that precede that byte byte are not hashed
  ```

Reply:
Updated again, please check new wording